components/platform/ws2812.c

```diff
@@ -32,112 +32,180 @@
-};
-
-// This is one eighth of 512 * 100ns, ie in total a bit above the requisite 50us
-const rmt_item32_t ws2812_rmt_reset = { .level0 = 0, .duration0 = 32, .level1 = 0, .duration1 = 32 };
 
 // descriptor for a ws2812 chain
 typedef struct {
   bool valid;
   bool needs_reset;
   uint8_t gpio;
+  rmt_item32_t reset;
+  rmt_item32_t bits[2];
   const uint8_t *data;
   size_t len;
+  uint8_t bitpos;
 } ws2812_chain_t;
 
 // chain descriptor array
 static ws2812_chain_t ws2812_chains[RMT_CHANNEL_MAX];
 
-#define MIN(a, b) ((a) < (b) ? (a) : (b))
-
 static void ws2812_sample_to_rmt(const void *src, rmt_item32_t *dest, size_t src_size, size_t wanted_num, size_t *translated_size, size_t *item_num)
 {
-  // Note: enabling these commented-out logs will ruin the timing so nothing
-  // will actually work when they're enabled. But I've kept them in as comments
-  // because they were useful in debugging the buffer management.
-  // ESP_DRAM_LOGW("ws2812", "ws2812_sample_to_rmt wanted=%u src_size=%u", wanted_num, src_size);
-
-  void *ctx;
-  rmt_translator_get_context(item_num, &ctx);
-  ws2812_chain_t *chain = (ws2812_chain_t *)ctx;
-
-  size_t reset_num = 0;
-  if (chain->needs_reset) {
-    // Haven't sent reset yet
-
-    // We split the reset into 8 even though it would fit in a single
-    // rmt_item32_t, simply so that dest stays 8-item aligned which means we
-    // don't have to worry about having to split a byte of src across multiple
-    // blocks (assuming the static asserts at the top of this file are true).
-    for (int i = 0; i < 8; i++) {
-      dest[i] = ws2812_rmt_reset;
-    }
-    dest += 8;
-    wanted_num -= 8;
-    reset_num = 8;
-    chain->needs_reset = false;
-  }
-
-  // Now write the actual data from src
-  const uint8_t *data = (const uint8_t *)src;
-  size_t data_num = MIN(wanted_num, src_size * 8) / 8;
-  for (size_t idx = 0; idx < data_num; idx++) {
-    uint8_t byte = data[idx];
-    for (uint8_t i = 0; i < 8; i++) {
-      dest[idx * 8 + i] = (byte & 0x80) ? ws2812_rmt_bit1 : ws2812_rmt_bit0;
-      byte <<= 1;
+  size_t cnt_in;
+  size_t cnt_out;
+  const uint8_t *pucData;
+  uint8_t ucData;
+  uint8_t ucBitPos;
+  esp_err_t tStatus;
+  void *pvContext;
+  ws2812_chain_t *ptContext;
+  uint8_t ucBit;
+
+  cnt_in = 0;
+  cnt_out = 0;
+  if( dest!=NULL && wanted_num>0 )
+  {
+    tStatus = rmt_translator_get_context(item_num, &pvContext);
+    if( tStatus==ESP_OK )
+    {
+      ptContext = (ws2812_chain_t *)pvContext;
+
+      if( ptContext->needs_reset==true )
+      {
+        dest[cnt_out++] = ptContext->reset;
+        ptContext->needs_reset = false;
+      }
+      if( src!=NULL && src_size>0 )
+      {
+        ucBitPos = ptContext->bitpos;
+
+        /* Each bit of the input data is converted into one RMT item. */
+
+        pucData = (const uint8_t*)src;
+        /* Get the current byte. */
+        ucData = pucData[cnt_in] << ucBitPos;
+
+        while( cnt_in<src_size && cnt_out<wanted_num )
```

Comment:
I suspect that there is an issue here if the `cnt_out<wanted_num` fails when not on a byte boundary. The original code validated that the entire byte would fit before it tried to encode it. There isn't a way to signal back to the rmt library that only part of a byte has been encoded.

Reply:
There is a good way to signal back that a byte was not completely processed: do not increase ```translated_size```.
The output parameter ```translated_size``` of the translator function reports how many bytes were processed. If a byte was not completely processed, it should not be considered in ```translated_size```. Additional information for the decoding process like the bit position can be stored in the translator context. Please note that the number of consumed raw bytes and the number of produced ```rmt_item32_t``` elements are not forced to have a fixed relationship of 1:8. The documentation has a note that it should be like this, but it is not a strict requirement (see [here](https://docs.espressif.com/projects/esp-idf/en/v4.3/esp32/api-reference/peripherals/rmt.html#_CPPv415sample_to_rmt_t)). The current routine inserts a reset signal with 8 ```rmt_item32_t``` elements without consuming a single raw byte for it (see [here](https://github.com/nodemcu/nodemcu-firmware/blob/10e378394861b8261eac4366658dad069aa12f63/components/platform/ws2812.c#L100)), and it is working apparently well.

I'm afraid my commit messages did not give a good explanation why I would like to change an implementation which is doing a good job. Playing around with the translator callback I found out that the transmission of RMT pulses stops if the ```wanted_num``` parameter is not met.

Example: I want to send 12 raw bytes which would give 96 ```rmt_item32_t``` elements after translation. The first call to the translator function has ```wanted_num``` set to 48, but my test code decides to provide only 40. The transmission stops after these pulses. The translator function is never called for the remaining 56 bytes of raw data. This was tested with an ESP32-C3.

Fortunately all observed values for the ```wanted_num``` parameter were multiples of 8. But I did not find a guarantee for it. That's why I changed the translator function. Am I too paranoid?

Reply:
Thanks -- I missed the saving of the bitpos in the context.